src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -84,22 +91,68 @@ public Http2Stream(HttpRequestMessage request, Http2Connection connection, int s
 
             public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
             {
-                // TODO: ISSUE 31312: Expect: 100-continue and early response handling
-                // Note that in an "early response" scenario, where we get a response before we've finished sending the request body
-                // (either with a 100-continue that timed out, or without 100-continue),
-                // we can stop send a RST_STREAM on the request stream and stop sending the request without tearing down the entire connection.
-
                 // Send request body, if any
                 if (_request.Content != null)
                 {
-                    using (Http2WriteStream writeStream = new Http2WriteStream(this))
+                    try
+                    {
+                        using (Http2WriteStream writeStream = new Http2WriteStream(this))
+                        {
+                            await _request.Content.CopyToAsync(writeStream, null, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        // Don't wait for completion, which could happen asynchronously.
+                        _ = _connection.SendEndStreamAsync(_streamId);
+                    }
+                    catch (Exception e)
                     {
-                        await _request.Content.CopyToAsync(writeStream, null, cancellationToken).ConfigureAwait(false);
+                         // We did not finish sending body so notify server.
+                         _ = _connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel);
+
+                        // if we decided abandon sending request and we get ObjectDisposed as result of it, just eat exception.
+                        if (_shouldSendRequestBody || (!(e is ObjectDisposedException) && !(e.InnerException is ObjectDisposedException)))
```

Comment:
How do we know that the ObjectDisposedExceptions we're eating here are being of abandoning sending a request?

Reply:
_shouldSendRequestBody be set to false when we get deny (300+) response code. Originally that was done only for 100Continue but it got changed. Assuming that sending more body will not change server's decision